File: algorithms/[1234]替换子串得到平衡字符串/Solution.java
// topics=["哈希表","滑动窗口"]

import java.util.HashMap;
import java.util.Map;

class Solution {
    /**
     * time O(|Σ|·n), space o(|Σ|), n 为 str 长度，|Σ| = 4
     */
    public int balancedString(String s) {
        int n = s.length();

        // map 对滑动窗口外的字符进行计数
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (map.containsKey(ch)) {
                map.put(ch, map.get(ch) + 1);
            } else {
                map.put(ch, 1);
            }
        }

        if (check(map, n / 4)) {
            return 0;
        }

        int i = 0, j = 0, res = n;
        while (j < n) {
            char right = s.charAt(j);
            map.put(right, map.get(right) - 1);
            j += 1;
            // 当滑动窗口外的字符均不超过平均数时，[i, j) 为符合要求的子串
            while (check(map, n / 4)) {
                res = Math.min(res, j - i);
                char left = s.charAt(i);
                map.put(left, map.get(left) + 1);
                i += 1;
            }
        }

        return res;
    }

    public boolean check(Map<Character, Integer> map, int avg) {
        if (map.containsKey('Q') && map.get('Q') > avg) {
            return false;
        }
        if (map.containsKey('W') && map.get('W') > avg) {
            return false;
        }
        if (map.containsKey('E') && map.get('E') > avg) {
            return false;
        }
        if (map.containsKey('R') && map.get('R') > avg) {
            return false;
        }
        return true;
    }
}

File: algorithms/[1234]替换子串得到平衡字符串/solution.py
# topics = ["哈希表", "滑动窗口"]

from collections import Counter


class Solution:
    def balancedString(self, s: str) -> int:
        """
        time O(|Σ|·n), space o(|Σ|), n 为 str 长度，|Σ| = 4
        """
        n = len(s)

        # target 对超出平均数的字符进行计数
        target = Counter(s)
        for ch, cnt in target.items():
            target[ch] = max(cnt - n // 4, 0)

        if target.total() == 0:
            return 0

        # current 对滑动窗口中的字符进行计数
        current: Counter[str] = Counter()
        i = j = 0
        res = n

        while j < n:
            current[s[j]] += 1
            j += 1
            # 当 current 与 target 相等，或者是后者的超集时，[i, j) 为符合要求的子串
            while current >= target:
                res = min(res, j - i)
                current[s[i]] -= 1
                i += 1

        return res
